# Vector_Database No.4 : [ PineCone_DB ] 

# Today Agenda :

- we will be going to talk about the "Vector Database".

- 4.PineCone_DB
- 5.Waviet_DB

- Pincone again a self managed database.
- Google it Pinecone Database URL : [ https://www.pinecone.io/ ]
- Almost similar to "Chrome_DB", all the vector Databases is almost same.

# Note :

- "Pinecone" is another vector database, a manage database which is going to provide complete indutry grade application building
- fesilities.
- Here home page,if we scoll bit down then, we can see they are providing a 
- Start Building :
- 1.Start building RAG Q & A
- 2.Multi model search
- 3.Semantic Search

- I don't have to go and Pay PineCode to this building this things, I can go and build without paying to a Pinecone, this is what 
- we will going to do here. Just to make us easy they are providing to us alots of examples.

# Note :

- Pinecone is another Vector Database, we can try to store the data inside Pinecode. 
- How to store a data inside a Pinecode.

- Let's understand :
- click on -> Database -> Indexes -> Here we can find installation we have to do.

# Libaries :

- Pinecode <- [ pip install pinecone ]

In [1]:
pip install pinecone

Note: you may need to restart the kernel to use updated packages.Collecting pinecone
   ---------------------------------------- 0.0/587.6 kB ? eta -:--:--
   ----------------- ---------------------- 262.1/587.6 kB ? eta -:--:--
   ----------------- ---------------------- 262.1/587.6 kB ? eta -:--:--
   ---------------------------------------- 587.6/587.6 kB 693.9 kB/s  0:00:00

   ---------------------------------------- 0/4 [pinecone-plugin-interface]
  Attempting uninstall: packaging
   ---------------------------------------- 0/4 [pinecone-plugin-interface]
    Found existing installation: packaging 25.0
   ---------------------------------------- 0/4 [pinecone-plugin-interface]
    Uninstalling packaging-25.0:
   ---------------------------------------- 0/4 [pinecone-plugin-interface]
      Successfully uninstalled packaging-25.0
   ---------------------------------------- 0/4 [pinecone-plugin-interface]
   ---------- ----------------------------- 1/4 [packaging]
   ---------- ---

# Next :

- once libaries install, we have to established the connectivity, with my local system and with Pinecone.
- Now  here just below -> Welcome to your project, click on -> (" Create index")
- URL : https://app.pinecone.io/organizations/-OcRp1riljMxNjdHl080/projects/d4a5e9ae-b49b-49c7-8b94-34c0d2c9bcf9/create-index/serverless
- So I will name it as -> ["vijaystory"], Need to select the model, there 3 options is showing to me to select.

- 1.LLama_text-embed-v2
- 2.Multilingual-e5-large
- 3.piecone-sparse-english-v0
- 4.text-embedding-3-large 
- 5.text-embedding-3-small <- This model we already have 
- 6.Manual configuration <- They are providing even this manual one as well.

- Bydefault, they are providing us the some of the embedding LLMs. This senario, we dont have to use any kind of LLMs locally.
- For example  :
- We are using -> 5.text-embedding-3-small <- which is avaiable with respect to the "EuriAPI".

- So, here out of these many option, which is avaible to me, I dont have to go and I dont have to convert anything into a embedding.
- in my local. I can simpley send, a data and automatically it will try to convert. But it is going to give me charges.
- Beginning, they are providing some where free small task, beyond that, they start charging you.
- 

# Next :

- But here, I dont want to use there model, I wanted to use my own, embedding models, Maybe from the ("Hugging Face") or 
- Maybe from our own EuriAPI, provided model, I would like to use that model, in that situation, we see here right corner 
- here list of all models, If I dont want to use it, we can see right corner there is -> "Custom setting" <- Click on this.
- I already named my index as a -> "vijaystory" now once I select this "Custom setting".

- Vector Type : Here 2 option 1.Dense | 2.Sparse 

- 1.Dense <- It means for every vector, we will be able to get a Non-Zero values. Means there will be some numbers.
- 2.Sparse <- It means a vector with the zero numbers. For example -> if we are going to use a Hot-Encoding. So Majority of time, vector 
- value will be the -> 00000000000000. That is something is called as Sparse.
- 
- So, Obviously, I have numeric so, I will select option, ("Dense")
- 1."Dimention" : of my model will going to -> [1536] <- OpenAI text-small model.
- 2."Metric" : Then whenever, I am going to store a embedding, whihc mechanismI would like to use for for "Distance" or for a Matching.
- So, I can use 3 option is here -> ( cosine/Euclidean/dotproduct ) <-

- Bydefault, it will be the "Cosine", otherwise, "Chroma_DB" also going to provide all these 3 options.
- "Faiss" : "Facebook AI Similarity Search" also providing to us all these 3 options. -> ( cosine/Euclidean/dotproduct ) <-
- All vector database is eventually providing all of these 3 options.

# Next :

- 1.Dense vector : It means, where we get the numbers -> [ -0.31,0.002,0.123 ] <- So, if we have this kind of vector, here we can see
- we have very less Zeros.
- 2.Sparse vector : It means, [0,0,0,0,0,1,0.01,0,0,0,0]<- Where we have majority of the data avaiable as a Zeros value is called 
- as Sparse vector.
- Now, Incase of Dense Vector, we will always end of getting a dimension, maybe a between a range of I would say, 256 to 4096.
- SO, here we can see Dimension of (1536) , if is a "Dense vector".
- If it is a Sparse vector, again a dimension is going to be very very high.

- If we remember in our childhood, we do a one-Hot-Encoding(OHE), incase one-hot-encoding, lets say we have a 10,000 unique words.
- in that case there will be a 10,000 zeros & out of those 10,000 zeros, only one is going to be one positive, means up value or one
- we can say, & rest of 9,999 values are going to be [0,0,0,0,0,0] or 00000000, it can go any kind of limits, depends upon machanism
- that we are trying to use, to convert our dataset. 
- So incase of dense, dimension will going to be very very compact. 
- Incase of Sparse, our dimension will going to be very very high.
- Again interms of capturing a semantic meaning, dense will be able to capture a Semantic meaning better way.
- Whereas, if I will talk about the "Sparse vector" it actually try to count the  frequency, means how many times the same data has
- occured into the entire document, which is not good interms of capturing a semantic meaning.
- So, Basically when I am trying to use a "RAG" kind of application wherever I am trying to build a recommendation system, wherever,
- I am trying to build this kind of semantic search operation. So, I could always use a "dense vector" for any kind of data.
- Not a "Sparse vector" one.

# Next : Continue....

- Now, 
- 1.Capacity mode : Serverless as it
- 2.Cloud provided : Bydefault cloud provider would be AWS.
- 3.Region : bydefault, I will select (Virginia us-east-1), I will pick
- Now, just create a "index".
- Now, once we create the index, we will get the URL :   [ https://vijaystory-8yfjwur.svc.aped-4627-b74a.pinecone.io ]
- Metric : cosine
- Dimension : 1536
- Host : URL : [ https://vijaystory-8yfjwur.svc.aped-4627-b74a.pinecone.io ]
- Cloud: AWS
- Region : us-east-1
- Type : Dense 
- Capacity mode : Serverless

- So, the URL, through which are able to connect, and My ["Vijaystory"] is up now.
- Now, I have to establish the connectivity, Connectivity with this URL :  [ https://vijaystory-8yfjwur.svc.aped-4627-b74a.pinecone.io ]
- which means, I have to basically hit the "Pinecone" database.
- Now, in right corner, we can see there is optionc all ("Connect"), once I click there, it providing me multiple language is there
- I will pick "Python" and I already installed the libary. So, I will copy this code and paste below we can see.
- So, here I have to replace my "API key" this  Asterisk (*) sybols section. so that it will hit the "Pinecone" server. The indexes which
- we have already craeted.
- From where, I will get the "APIKey"- same URL of
- "Pinecone", URL : [https://app.pinecone.io/organizations/-OcRp1riljMxNjdHl080/projects/d4a5e9ae-b49b-49c7-8b94-34c0d2c9bcf9/keys]
- there is "API Key" click there, create "APIKey", I will name it as ["PineconeAPI"]
- MY API key -> [" pcsk_2UjGDB_QQaKPyYx3NJcyTnQ65QZcSGEJNJ6MJXr7QHeM216kCPg9VANgxiJSBKXdkCAWej "]

In [1]:
# code below copied from the pinecone

# from pinecone import Pinecone

# pc = Pinecone(api_key="********-****-****-****-************")
# index = pc.Index("quickstart")

# Timing : 7:41 PM

- Now my code is ready, I replace with my Asterisk (*) with my orginal "API key" which I have just, created.
- 2nd changes, I made was I replace default name -> ("quickstart") to my created index name as -> ["vijaystory"] &
- Now execute the code.

In [1]:
from pinecone import Pinecone

pc = Pinecone(api_key="pcsk_2UjGDB_QQaKPyYx3NJcyTnQ65QZcSGEJNJ6MJXr7QHeM216kCPg9VANgxiJSBKXdkCAWej") 
index = pc.Index("vijaystory")

c:\Users\vijay\.conda\envs\venv\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


# Next :

- Now, once I am able to established a connect, with respect to "Pinecone".
- Below is the same data, I am going to use it, which I have used previously also.

In [2]:
texts = [
    "Sudhanshu Kumar was born in Jamshedpur, Jharkhand, India, to a modest family.",
    "His early life was marked by financial hardship and limited resources.",
    "He believed education could transform his life and pursued it relentlessly.",
    "Sudhanshu earned a Computer Science and Engineering degree.",
    "He worked at Wipro, Deloitte, Verizon Labs, and Ernst & Young.",
    "He gained expertise in SAP WebDynpro, Fiori UI5 HANA, Java, Big Data, and Data Analytics.",
    "Sudhanshu remained committed to making education accessible for everyone.",
    "His teaching empowered countless students to build their careers.",
    "Many students credit him with changing their lives through affordable learning.",
    "His journey is a testament to triumph over adversity and the power of knowledge."
]

# Generate the Embedding :

- To generate the embedding, also I am going to use the same function, previous one.

In [3]:
import requests
import numpy as np

def generate_embeddings(text):
    url = "https://api.euron.one/api/v1/euri/embeddings"
    headers = {
        "Content-Type": "application/json",
        "Authorization": "Bearer euri-46d7971d2ffa9e1c356959de86118e7c51ab8390c3d8608f3f4d545cde21c0a3"
    }
    payload = {
        "input": text,
        "model": "text-embedding-3-small"
    }

    response = requests.post(url, headers=headers, json=payload)
    data = response.json()
    
    embedding = np.array(data['data'][0]['embedding'])
    
    return embedding

- Even For Loop wise, I am going to use the same. Just execute then now it created the embeddings.

In [4]:
embeddings = [generate_embeddings(i).tolist() for i in texts]

In [5]:
embeddings

[[-0.024888895,
  -0.021206673,
  0.0047149123,
  0.018578006,
  0.0033718925,
  0.0052990606,
  0.023324212,
  0.029666394,
  0.0044854255,
  -0.033859745,
  0.008266743,
  -0.03031313,
  -0.017284535,
  -0.015991064,
  0.033400774,
  0.014541123,
  0.025598219,
  -0.06951366,
  -0.00032141202,
  0.012527897,
  0.07581412,
  0.0011356992,
  0.02186384,
  -0.015865888,
  0.0368848,
  -0.013956975,
  -0.053240955,
  -0.039888993,
  -0.014176031,
  -0.04994469,
  -0.023011275,
  0.0015972808,
  0.0042011747,
  -0.01134917,
  0.025034932,
  -0.019193448,
  -0.00964888,
  0.010702434,
  -0.0069941343,
  -0.0058362684,
  -0.019579403,
  -0.05908244,
  -0.02280265,
  0.04548013,
  -0.018317226,
  0.020080103,
  -0.0056902315,
  0.02518097,
  -0.04969434,
  -0.0054190196,
  -0.00793816,
  -0.020132259,
  0.061919734,
  -0.00044854253,
  0.008798735,
  -0.008715286,
  0.01570942,
  0.0278305,
  -0.0042272527,
  -0.039304845,
  0.0012843442,
  -0.018817924,
  0.013602314,
  0.003781318,
  0.007

# Next :

- Now, once Embeddings are created, we can try to store this data, even till this point, we have done the exact same thing with respect 
- "Chroma_DB", even with the "Faiss_DB", even with "Qdrant_DB", we have done the same thing.
- I am able to convert my data into Embedding. 

# Next :

- Here next is insert the data, insert a record, to insert the record my entry point should be ["index"].
- The ["index"], which I have been created as an object. ["vijaystory"] this one -> [ index = pc.Index("vijaystory") ]
- I can try to call basically, "index" & there is function called as ("absurd"). So, ("absurd"), which is avaiable for "Pinecone".
- Then, we can try to store a data, may be one by one by one or may be I can try to store all the data in one go, that is also possible.
- one single go record, I can try to store and one single embedding, I can try to store, or I can try to store whole dataset. 
- These all are possible, I can do that, so, here we are going to store whole data.


# Note :

- Let's start with "single Storage" 1st, there is parameter called as vector, now inside that, we can try to write whatever I would like
- to store over here, 

- Code Line No.1 : [ index.upsert ]

- insert a record, to insert the record my entry point should be ["index"].
- The ["index"], which I have been created as an object. ["vijaystory"] this one -> [ index = pc.Index("vijaystory") ]
- I can try to call basically, "index" & there is function called as ("absurd"). So, ("absurd"), which is avaiable for "Pinecone".


- Code Line No.2 : [ Vectors ]

- There is a parameter called as "Vector", inside that we can try to maybe like bright, whatever I would like store over here, so, inside 
- this vector, I can try to give "id" = [0].

- Code Line No.3 : str(0)
- I am try to give "id" : str(0), the same which we have given for "Chome_DB". So, I am giving basically "id" = [0]

- Code Line No.4 : Embeddings[0]
- I am trying to call basically a Embeddings, as [0], it basically means, I am going to store a [0th] record from a embeddings vector.

- Code Line No.5 : ["text" : texts[0]]
- I am going to store only one record, & text is again Zero. Means again for the same, I am going to store that "text information" 
- means a complete line, I am going to store over here.

# Outcome or Output : 

- As a outcome, we can see that -> {'upserted_count': 1}, just go back on "pincone" and Go to "Database",
- click on "Index" which is -> "vijaystory", scroll below then, we can see that "id": 0
- text: "Sudhanshu Kumar was born in Jamshedpur, Jharkhand, India, to a modest family." <- This is the record, which I am going to see.
- Now, If I will click three dot (:) of text : "Sudhanshu Kumar was born in Jamshedpur, Jharkhand, India, to a modest family."  
- Now same line, click on 3 dot(:) and "edit" pop up open, here I can see the ID : [0], this is bydefault, which we have given.

- "Dense Value" or "Dense Vector" <- this is the vector
- this is the embeddings -> [ -0.0254169162,-0.0186557658,0.00634901226,0.0197826251,0.00278323586,0.00340666,0.0245404709,etc........],
- it is able to store.
- This is the "text" : [ Sudhanshu Kumar was born in Jamshedpur, Jharkhand, India, to a modest family ] <- which it has stored.
- "Metadata" : (?) <- we have not added anything, If I would like to add I can add that is completely ok.
- So, all of these values are recorded inside my Database.
- So, we are able to store our very 1st record which is -> [0] <-.

In [7]:
index.upsert(
    vectors = [(
        str(0),
        embeddings[0],
        {"text": texts[0]}
    )])

{'upserted_count': 1}

- Emebeddings[0] <- which means, this one the outcome we can here -> embeddings[0] &

- For which data?
- for this data -> "texts[0]" -> ['Sudhanshu Kumar was born in Jamshedpur, Jharkhand, India, to a modest family.'], I am trying to
- store the lines.
- then, I am trying to store the [ "embeddings[0]" ] & 
- I am trying to store the ids, which is ["str(0)"]
- This much, I am trying to store over here.
- This is how, I can store one records.

In [8]:
embeddings[0]

[-0.025416916,
 -0.018655766,
 0.0063490123,
 0.019782625,
 0.0027832359,
 0.00340666,
 0.02454047,
 0.027628899,
 0.007100251,
 -0.03418137,
 0.007533257,
 -0.031927653,
 -0.015546472,
 -0.016089033,
 0.031739846,
 0.014941308,
 0.02835927,
 -0.07128423,
 -0.0006270106,
 0.012030257,
 0.07570819,
 0.00065244315,
 0.022015475,
 -0.016798537,
 0.037269797,
 -0.014210937,
 -0.05383879,
 -0.039481778,
 -0.017403701,
 -0.048580118,
 -0.024206588,
 0.0020346055,
 0.0016224674,
 -0.0109033985,
 0.025270842,
 -0.018509692,
 -0.0067924517,
 0.011237282,
 -0.007183722,
 -0.0059838267,
 -0.019626116,
 -0.059473082,
 -0.021785928,
 0.046827227,
 -0.018947914,
 0.022161549,
 -0.0051491167,
 0.02547952,
 -0.05049995,
 -0.00524563,
 -0.007955829,
 -0.021076426,
 0.05922267,
 -0.0011496824,
 0.007037648,
 -0.007376749,
 0.017539343,
 0.027858444,
 -0.006067298,
 -0.03689418,
 0.00024291362,
 -0.020022605,
 0.015473435,
 0.0046117725,
 0.007820188,
 -0.04336318,
 0.016297711,
 -0.01844709,
 -0.0461594

In [6]:
texts[0]

'Sudhanshu Kumar was born in Jamshedpur, Jharkhand, India, to a modest family.'

# Store Multiple record :

- If, I have to store multiple records, I can try to run it through a "loop" and Then I will be able to store the entire record.

# Public Question :

- What is "Meta data" visually do?
- So, any kind of a key : Value pair, maybe we can try to add some sort of a label. So, file name & its value and maybe page number.
- "Meta Data" or "Meta Information" is basically means nothing data of data, information about the data, that depends upon me.

- Now lets suppose, I want to store all the records, that case, I have to create function something like above we have written for the
- single records.
- so, one by one, one by one, I can try to store the complete values.

# Next :

- Now, I would like to go ahead with the entire dataset. before that lets create a "record" variable 1st.
- Lets say, I am going to create a 

- Code NO.1 : record []
- Lets suppose, I am creating one single, "List" over here.
- Now I will try to create a data in above format, which we did above part for single record. 

- Code No.2 : [ for i in range(len(texts)) : ]
- for i in range of this one -> len(texts).

- Code No.3 : [ records.append((str(i), embeddings[i], {"text": texts[i]}) ]

- So record.append 
- 1."ids" which is (str(i))
- 2.embeddings &
- 3."text": texts[i] <- then a data, a textual information.

# Outcome :

- Now, If I will call the "record", as we can as a outcome, we are getting 3 things 
- 1.Ids, which is [0]
- 2.embedding : this one -> [-0.025416916, -0.018655766, 0.0063490123, 0.019782625,] 
- 3.Data : -> {'text': 'His journey is a testament to triumph over adversity and the power of knowledge.'}
- So, Yes we are able to convert this entire data into a records.

In [11]:
records = []
for i in range(len(texts)):
    records.append((str(i), embeddings[i], {"text": texts[i]}))

In [12]:
records

[('0',
  [-0.025416916,
   -0.018655766,
   0.0063490123,
   0.019782625,
   0.0027832359,
   0.00340666,
   0.02454047,
   0.027628899,
   0.007100251,
   -0.03418137,
   0.007533257,
   -0.031927653,
   -0.015546472,
   -0.016089033,
   0.031739846,
   0.014941308,
   0.02835927,
   -0.07128423,
   -0.0006270106,
   0.012030257,
   0.07570819,
   0.00065244315,
   0.022015475,
   -0.016798537,
   0.037269797,
   -0.014210937,
   -0.05383879,
   -0.039481778,
   -0.017403701,
   -0.048580118,
   -0.024206588,
   0.0020346055,
   0.0016224674,
   -0.0109033985,
   0.025270842,
   -0.018509692,
   -0.0067924517,
   0.011237282,
   -0.007183722,
   -0.0059838267,
   -0.019626116,
   -0.059473082,
   -0.021785928,
   0.046827227,
   -0.018947914,
   0.022161549,
   -0.0051491167,
   0.02547952,
   -0.05049995,
   -0.00524563,
   -0.007955829,
   -0.021076426,
   0.05922267,
   -0.0011496824,
   0.007037648,
   -0.007376749,
   0.017539343,
   0.027858444,
   -0.006067298,
   -0.03689418,


# Note :

- Now, I have all the records, now I have to update this record in one single sort inside my Database.
- So, I can try to call, "index.upsert" & then Vectors=records.
- because the records are already avaiable into a compatiable formt. 
- If we execute this below code. In a one single go, it is able to insert all the records.
- As a outcome, we can see that -> {'upserted_count': 10} <- So, total 10 records it is able to Upserts.
- Now, go to the "Pinecone_DB" website and refresh it "indexes"  then we are able to see that, it has updated the all the 10 records.
- URL:https://app.pinecone.io/organizations/-OcRp1riljMxNjdHl080/projects/d4a5e9ae-b49b-49c7-8b94-34c0d2c9bcf9/indexes/vijaystory/browser
- So, this is how we can insert the data, is already avialable inside my database.

In [14]:
index.upsert(vectors=records)

{'upserted_count': 10}

# Next : "Search operation"

- Once all my records are avialable inside Database, now we can go-ahead with the "Search operation"
- for this, I can try to create a "Query", my query is -> [" Sudhanshu has worked at Wipro" ] <-

# Steps : Need to do is.

- 1st things is that, we have to convert the "Query" into a embeddings, Once my query is available into an embedding.
- I can go-ahead and do the search.

In [7]:
query = "sudhanshu has worked at wipro"

- To convert this query into the embedding, I can call simply call, my generate_embeddings function &
- "query.tolist()" 
- As a outcome, we can see that it is able to generate an embedding for this query for this line. -> ["sudhanshu has worked at wipro"]
- it is able to generate the embeddings, now I have just need to do "Cosine Similarity".

- Let me keep inside the variable as [ "query_embed" ]. 
- So, my this query_embedding -> [ "sudhanshu has worked at wipro" ] <- is aviable into the numerical space.

In [8]:
query_embed = generate_embeddings("query").tolist()

In [9]:
query_embed

[-0.008969661,
 0.0370767,
 0.005726238,
 -0.012894873,
 -0.020335434,
 0.020083213,
 0.0037951597,
 0.04499018,
 0.0020374844,
 -0.030061763,
 -0.00848098,
 -0.04713407,
 -0.007160753,
 0.006297679,
 0.0051942063,
 0.013572721,
 -0.037139755,
 -0.017560989,
 0.028721832,
 0.062141307,
 0.023961132,
 0.0046345876,
 -0.0134072,
 -0.020713769,
 -0.007176517,
 0.027807524,
 -0.033419475,
 -0.0041971393,
 0.022400506,
 -0.022148283,
 0.0069203535,
 -0.05202088,
 0.051358797,
 -0.044075873,
 0.022463562,
 -0.015243694,
 0.038306285,
 0.017198417,
 -0.030849958,
 -0.022904951,
 0.0043508373,
 -0.014565846,
 0.001260127,
 0.041616704,
 0.023314813,
 0.03206378,
 -0.03906295,
 -0.0340185,
 0.039409757,
 0.015251576,
 -0.01401411,
 -0.02523801,
 -0.0031626332,
 0.1640707,
 -0.00127195,
 -0.0135175465,
 -0.021596547,
 0.030282458,
 0.019626059,
 0.007657316,
 -0.029131692,
 -0.02031967,
 0.054700743,
 0.06415909,
 -0.0082602855,
 -0.012918519,
 -0.03477517,
 0.045400042,
 0.006683895,
 -0.008748

# Next :

- Same query which, I am using last time everything is same, expect the connectivity.

- Now once, I am able to convert my query into the embeddings, I can simply try to call.
- Index.query, so default query is already avaiable.
- So, index.query, then I can try to pass -> ["query_embed"] vector <- Then I can try to pass Top-K search ->[ top_K=3 ] or top 5 or top 6
- whatever I want to provide, & then include the "Meta data" ["include_Metadata=True"]<- we can keep True or False.

# Output :

- Here, as a outcome we can that it is able to give me the matches.
- The very 1st match is -> matches': ['id': '4'] <- id=4, or interm computer term [4] means, which is 5th document technically 
- because start with [0,1,2,3,4].
- what is the "Cosin distance", It is going to be ['score': 0.133443832 ] or 0.13 %. | ['score': 0.11658287] or 0.11% |
- ['score': 0.0967788696] or 0.96%.
- As we can -> ['score': 0.11658287] or 0.11% <- this one is nearest one.
- Then 2nd nearest one is -> ['score': 0.133443832 ] or 0.13 % <-

- So, -> [ Id:4, Id:5 & id:9 & then Id:0] <- it is able to give me. The final answer, which I am able to received, interms of 
- "Cosin Similaarity" distance check.

In [10]:
index.query(vector=query_embed, top_k=3, include_metadata=True)

{'matches': [{'id': '4',
              'metadata': {'text': 'He worked at Wipro, Deloitte, Verizon '
                                   'Labs, and Ernst & Young.'},
              'score': 0.133443832,
              'values': []},
             {'id': '5',
              'metadata': {'text': 'He gained expertise in SAP WebDynpro, '
                                   'Fiori UI5 HANA, Java, Big Data, and Data '
                                   'Analytics.'},
              'score': 0.11658287,
              'values': []},
             {'id': '9',
              'metadata': {'text': 'His journey is a testament to triumph over '
                                   'adversity and the power of knowledge.'},
              'score': 0.0967788696,
              'values': []}],
 'namespace': '',
 'usage': {'read_units': 1}}